File: llm-server/r2_client.py
"""
R2 Client — Cloudflare R2 storage via boto3 S3 API.

Stores project files at: projects/{project_id}/{relative_path}
Metadata at: projects/{project_id}/_meta.json
"""

import json
import os
import logging
from typing import Optional

import boto3
from botocore.exceptions import ClientError

log = logging.getLogger("r2")

EXCLUDED_PREFIXES = [
    "node_modules/", ".git/", "dist/", ".next/",
    "build/", "__pycache__/", ".cache/", ".turbo/",
]

DEFAULT_WORKDIR = "/home/user/app/"


def _get_client():
    account_id = os.getenv("CF_ACCOUNT_ID")
    bucket = os.getenv("R2_BUCKET_NAME")
    access_key = os.getenv("R2_ACCESS_KEY_ID")
    secret_key = os.getenv("R2_SECRET_ACCESS_KEY")

    if not all([account_id, bucket, access_key, secret_key]):
        raise RuntimeError("CF_ACCOUNT_ID, R2_BUCKET_NAME, R2_ACCESS_KEY_ID, R2_SECRET_ACCESS_KEY required")

    client = boto3.client(
        "s3",
        endpoint_url=f"https://{account_id}.r2.cloudflarestorage.com",
        aws_access_key_id=access_key,
        aws_secret_access_key=secret_key,
        region_name="auto",
    )
    return client, bucket


def _is_excluded(path: str) -> bool:
    return any(path.startswith(p) for p in EXCLUDED_PREFIXES)


def _key(project_id: str, relative_path: str) -> str:
    return f"projects/{project_id}/{relative_path}"


# --- File operations ---

def r2_put(project_id: str, relative_path: str, content: str) -> None:
    if _is_excluded(relative_path):
        return
    client, bucket = _get_client()
    key = _key(project_id, relative_path)
    client.put_object(Bucket=bucket, Key=key, Body=content.encode("utf-8"))
    log.info(f"PUT {relative_path} ({len(content)} bytes)")


def r2_get(project_id: str, relative_path: str) -> Optional[str]:
    client, bucket = _get_client()
    key = _key(project_id, relative_path)
    try:
        resp = client.get_object(Bucket=bucket, Key=key)
        return resp["Body"].read().decode("utf-8")
    except ClientError as e:
        if e.response["Error"]["Code"] in ("NoSuchKey", "404"):
            return None
        raise


def r2_delete(project_id: str, relative_path: str) -> None:
    client, bucket = _get_client()
    key = _key(project_id, relative_path)
    client.delete_object(Bucket=bucket, Key=key)


def r2_list_files(project_id: str) -> list[str]:
    client, bucket = _get_client()
    prefix = f"projects/{project_id}/"
    files = []
    continuation_token = None

    while True:
        kwargs = {"Bucket": bucket, "Prefix": prefix, "MaxKeys": 1000}
        if continuation_token:
            kwargs["ContinuationToken"] = continuation_token

        resp = client.list_objects_v2(**kwargs)

        for obj in resp.get("Contents", []):
            rel = obj["Key"][len(prefix):]
            if rel and rel != "_meta.json":
                files.append(rel)

        if resp.get("IsTruncated"):
            continuation_token = resp.get("NextContinuationToken")
        else:
            break

    return files


# --- Metadata ---

def r2_put_meta(project_id: str, meta: dict) -> None:
    client, bucket = _get_client()
    key = f"projects/{project_id}/_meta.json"
    client.put_object(
        Bucket=bucket,
        Key=key,
        Body=json.dumps(meta).encode("utf-8"),
        ContentType="application/json",
    )


def r2_get_meta(project_id: str) -> Optional[dict]:
    client, bucket = _get_client()
    key = f"projects/{project_id}/_meta.json"
    try:
        resp = client.get_object(Bucket=bucket, Key=key)
        return json.loads(resp["Body"].read().decode("utf-8"))
    except ClientError as e:
        if e.response["Error"]["Code"] in ("NoSuchKey", "404"):
            return None
        raise


# --- Path helper ---

def to_relative_path(absolute_path: str, workdir: str = DEFAULT_WORKDIR) -> str:
    if absolute_path.startswith(workdir):
        return absolute_path[len(workdir):]
    if absolute_path.startswith("./"):
        return absolute_path[2:]
    return absolute_path

File: src/data/r2-client.ts
import { AwsClient } from 'aws4fetch'

// --- R2 S3-compatible API client ---

export type ProjectMeta = {
  templateName: string
  lastSync: number
  envVars?: Record<string, string>
}

const EXCLUDED_PREFIXES = [
  'node_modules/',
  '.git/',
  'dist/',
  '.next/',
  'build/',
  '__pycache__/',
  '.cache/',
  '.turbo/',
]

function isExcluded(path: string): boolean {
  return EXCLUDED_PREFIXES.some((p) => path.startsWith(p))
}

function getClient() {
  const accessKeyId = process.env.R2_ACCESS_KEY_ID
  const secretAccessKey = process.env.R2_SECRET_ACCESS_KEY
  const accountId = process.env.CF_ACCOUNT_ID
  const bucket = process.env.R2_BUCKET_NAME

  if (!accountId || !bucket) {
    throw new Error('CF_ACCOUNT_ID and R2_BUCKET_NAME are required')
  }

  if (!accessKeyId || !secretAccessKey) {
    throw new Error('R2_ACCESS_KEY_ID and R2_SECRET_ACCESS_KEY are required')
  }

  return {
    client: new AwsClient({ accessKeyId, secretAccessKey, service: 's3', region: 'auto' }),
    base: `https://${accountId}.r2.cloudflarestorage.com/${bucket}`,
  }
}

// --- File operations ---

export async function r2Put(projectId: string, relativePath: string, content: string): Promise<void> {
  if (isExcluded(relativePath)) return
  const { client, base } = getClient()
  const url = `${base}/projects/${projectId}/${relativePath}`
  console.log(`[r2] PUT ${relativePath} (${content.length} bytes) → ${url}`)
  const res = await client.fetch(url, {
    method: 'PUT',
    body: content,
  })
  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new Error(`R2 PUT failed (${res.status}): ${body}`)
  }
  console.log(`[r2] PUT ${relativePath} OK`)
}

export async function r2Get(projectId: string, relativePath: string): Promise<string | null> {
  const { client, base } = getClient()
  const res = await client.fetch(`${base}/projects/${projectId}/${relativePath}`)
  if (res.status === 404) return null
  if (!res.ok) throw new Error(`R2 GET failed (${res.status})`)
  return res.text()
}

export async function r2Delete(projectId: string, relativePath: string): Promise<void> {
  const { client, base } = getClient()
  await client.fetch(`${base}/projects/${projectId}/${relativePath}`, { method: 'DELETE' })
}

export async function r2ListFiles(projectId: string): Promise<string[]> {
  const { client, base } = getClient()
  const prefix = `projects/${projectId}/`
  const files: string[] = []
  let token: string | undefined

  do {
    const params = new URLSearchParams({ 'list-type': '2', prefix, 'max-keys': '1000' })
    if (token) params.set('continuation-token', token)

    const res = await client.fetch(`${base.split('/').slice(0, -1).join('/')}/${base.split('/').pop()}?${params}`)
    if (!res.ok) throw new Error(`R2 LIST failed (${res.status})`)

    const xml = await res.text()
    for (const m of xml.matchAll(/<Key>([^<]+)<\/Key>/g)) {
      const rel = m[1].slice(prefix.length)
      if (rel && rel !== '_meta.json') files.push(rel)
    }

    token = xml.includes('<IsTruncated>true')
      ? xml.match(/<NextContinuationToken>([^<]+)/)?.[1]
      : undefined
  } while (token)

  return files
}

// --- Metadata ---

export async function r2PutMeta(projectId: string, meta: ProjectMeta): Promise<void> {
  const { client, base } = getClient()
  await client.fetch(`${base}/projects/${projectId}/_meta.json`, {
    method: 'PUT',
    body: JSON.stringify(meta),
    headers: { 'Content-Type': 'application/json' },
  })
}

export async function r2GetMeta(projectId: string): Promise<ProjectMeta | null> {
  const { client, base } = getClient()
  const res = await client.fetch(`${base}/projects/${projectId}/_meta.json`)
  if (res.status === 404) return null
  if (!res.ok) throw new Error(`R2 GET meta failed (${res.status})`)
  return res.json() as Promise<ProjectMeta>
}

// --- Path helper ---

const DEFAULT_WORKDIR = '/home/user/app/'

export function toRelativePath(absolutePath: string, workdir = DEFAULT_WORKDIR): string {
  if (absolutePath.startsWith(workdir)) return absolutePath.slice(workdir.length)
  if (absolutePath.startsWith('./')) return absolutePath.slice(2)
  return absolutePath
}
